lib/tests/formulae_bottle.rb

```diff
@@ -0,0 +1,386 @@
+        if args.keep_old? && formula.requirements.exclude?(LinuxRequirement.new)
+          @testable_dependents = @bottled_dependents = @source_dependents = []
+          return
+        end
+
+        info_header "Determining dependents..."
+
+        build_dependents_from_source_allowlist = %w[
+          cabal-install
+          docbook-xsl
+          erlang
+          ghc
+          go
+          ocaml
+          ocaml-findlib
+          ocaml-num
+          openjdk
+          rust
+        ]
+
+        uses_args = %w[--formula --include-build --include-test]
+        uses_args << "--recursive" unless args.skip_recursive_dependents?
+        dependents = with_env(HOMEBREW_STDERR: "1") do
+          Utils.safe_popen_read("brew", "uses", *uses_args, formula_name)
+               .split("\n")
+        end
+        dependents -= @formulae
+        dependents = dependents.map { |d| Formulary.factory(d) }
+
+        dependents = dependents.zip(dependents.map do |f|
+          if args.skip_recursive_dependents?
+            f.deps
+          else
+            begin
+              f.recursive_dependencies
+            rescue TapFormulaUnavailableError => e
+              raise if e.tap.installed?
+
+              e.tap.clear_cache
+              safe_system "brew", "tap", e.tap.name
+              retry
+            end
+          end.reject(&:optional?)
+        end)
+
+        # Defer formulae which could be tested later
+        # i.e. formulae that also depend on something else yet to be built in this test run.
+        dependents.select! do |_, deps|
+          still_to_test = @formulae - @built_formulae
+          (deps.map { |d| d.to_formula.full_name } & still_to_test).empty?
+        end
+
+        # Split into dependents that we could potentially be building from source and those
+        # we should not. The criteria is that it depends on a formula in the allowlist and
+        # that formula has been, or will be, built in this test run.
+        @source_dependents, dependents = dependents.partition do |_, deps|
+          deps.any? do |d|
+            full_name = d.to_formula.full_name
+
+            next false unless build_dependents_from_source_allowlist.include?(full_name)
+
+            @formulae.include?(full_name)
+          end
+        end
+
+        # From the non-source list, get rid of any dependents we are only a build dependency to
+        dependents.select! do |_, deps|
+          deps.reject { |d| d.build? && !d.test? }
+              .map(&:to_formula)
+              .include?(formula)
+        end
+
+        dependents = dependents.transpose.first.to_a
+        @source_dependents = @source_dependents.transpose.first.to_a
+
+        @testable_dependents = @source_dependents.select(&:test_defined?)
+        @bottled_dependents = with_env(HOMEBREW_SKIP_OR_LATER_BOTTLES: "1") do
+          dependents.select(&:bottled?)
+        end
+        @testable_dependents += @bottled_dependents.select(&:test_defined?)
```

Comment:
Move all this to `FormulaeDependents` (necessary to access the variables anyway).